R/fragments-calculate2.R

```diff
@@ -0,0 +1,279 @@
+        aion <- character()
+    }
+    
+    ## fragment str (carboxyl-terminus)
+    ctype <- rep(c("x", "y", "z")[ctype], each = tn * zn)
+    if (length(ctype)) {
+        cion <- paste0(ctype, pos)
+    } else {
+        cion <- character()
+    }
+    
+    ## generate unique dataframe with all fragments and modifications 
+    for (i in 1:length(mod_combinations)) {
+        df[[i]] <- data.frame(mz = c(amz_mod[[i]], cmz_mod[[i]]),
+                              ion = c(aion, cion),
+                              type = c(atype, ctype),
+                              pos = pos,
+                              z = z,
+                              seq = c(aseq, cseq),
+                              stringsAsFactors = FALSE)
+        df[[i]] <- .neutralLoss(df[[i]],
+                                water = neutralLoss$water,
+                                ammonia = neutralLoss$ammonia)
+        df[[i]] <- .terminalModifications(df[[i]],
+                                          modifications = fixed_modifications)
+        rownames(df[[i]]) <- NULL
+        non_zero <- mod_combinations[[i]] != 0
+        names(mod_combinations[[i]])[non_zero] <- 
+            paste0("[", (names(mod_combinations[[i]])[non_zero]), "]")
+        df[[i]][["peptide"]] <- paste(names(mod_combinations[[i]]),
+                                      collapse = "")
+    }
+    
+    df <- do.call(rbind, df)
+    rownames(df) <- NULL
+    df
+}
+
+#' @title Generates list of possible combinations of modifications
+#' 
+#' @param sequence Character. A peptide sequence that may have modifications or not 
+#' 
+#' @param fixed_modifications Named numeric. Specifies which fixed modifications are used
+#' 
+#' @param variable_modifications Named numeric. Specifies which variable modifications are used
+#' 
+#' @param max_mods Numeric. Indicates how many modifications can be applied at once.
+#' 
+#' @return list with all possible combinations of modifications
+#' 
+#' @noRd
+#' 
+#' @examples
+#' .modificationPositions("ARGHKA", variable_modifications = c(A = 4, K = 5, S = 8), max_mods = 3)
+#' 
+.modificationPositions <- function(sequence,
+                                   variable_modifications = NULL,
+                                   max_mods = Inf) {
+    
+    sequence_split <- strsplit(sequence, "")[[1]]
+    
+    if (length(variable_modifications)) {
+        modifiable_positions_var <-
+            which(sequence_split %in% names(variable_modifications))
+        l <- length(modifiable_positions_var)
+    } else {l <- NULL}
+    
+    ## take the maximum amount of modifications possible 
+    max_mods <- min(max_mods, l)
+    
+    if (max_mods > 0 & length(l)) {
+        mod_combinations <- lapply(0:max_mods, function(n) {
+            if (n == 0) {
+                list(integer(0))
+            } else if (l == 1) {
+                list(modifiable_positions_var)
+            } else {
+                combn(modifiable_positions_var, n, simplify = FALSE)
```

Comment:
This `if/else if/else` statement is independent of the `n` in the loop/lapply. I would take it outside the loop.
e.g.
```r
    mod_combinations <- c(
        list(integer(0)),
        if (length(modifiable_positions_var) == 1)
            list(modifiable_positions_var)
        else
            unlist(
                lapply(seq_len(max_mods),
                    function(n)combn(
                        modifiable_positions_var, n, simplify = FALSE
                    )
                ),
                recursive = FALSE
            )
    )
```